### ISSUE TITLE
Support multiple accounts

### ISSUE THREAD

**Issue by varac** [MoritzR/fints2ledger#7]
Right now, the config file only handles one account (`selectedAccount`).
It would be great if you could specify multiple accounts in one config file in order to de-duplicate configuration (right now you would need 3 config files for 3 account i.e.)/

**Comment by varac** [comment]
I.e. I have two account which I use sometimes for similar stuff (i.e. shopping). That means I need to maintain two similar filter rule sets in two config files.

**Comment by varac** [comment]
The current state requires multiple config files for multiple accounts, so it would be good to be able to specify the config file to use as a cli argument.

**Comment by varac** [comment]
btw, [buchhaltung](https://github.com/dpaetzel/buchhaltung) handles multiple accounts like [this](https://github.com/dpaetzel/buchhaltung/blob/master/config.yml#L55)

**Comment by MoritzR** [comment]
I had a look at [buchhaltung](https://github.com/dpaetzel/buchhaltung) and that looks like a much more evolved version of `fints2ledger`.
I gave up trying it out after some hours of failing to set it up, though.
What's the reason you don't use this over `fints2ledger`?

**Comment by varac** [comment]
I have been using buchhaltung, I think I moved to fints2ledger becuase haskell is a complete black box to me, and I feel much better with a python project, where I can at least contribute/improve small things. And so far it works for me quite well. 
I have a wrapper script around fints2ledger to deal with multiple accounts, so we can also close this if it's not realistic to add native support.

**Comment by MoritzR** [comment]
I would like to keep this issue open.
It documents what is not supported, and I still hope I someday find the time and motivation to add this feature 🙂

**Comment by stewmehr** [comment]
I guess I am a little late to the party because I have only just discovered `fints2ledger` in its current Python form and now the move to Haskell is well underway. I am very much willing to contribute further to this project, because automated CSV downloads integrate very nicely with other tools like [hledger-flow](https://github.com/apauley/hledger-flow) (my current weapon of choice).

Two features that I would love to work on:
- native support for multiple accounts (as requested here by @varac)
- finer control regarding the transaction data fields to be downloaded. In particular, I am interested in the account balances associated with each booking for automated balance assertions (see relevant doc sections in [ledger](https://ledger-cli.org/doc/ledger3.html#Balance-assertions)/[hledger](https://hledger.org/1.29/hledger.html#balance-assertions)).

The only issue is: I, too, have zero prior experience with Haskell (and limited to intermediate Python skills tbh). Hence, I expect the learning curve to be quite steep initially. I have not checked out the Haskell branch yet but can I kindly ask you to map out a rough vision for this project (maybe in a separate issue), @MoritzR? Do you want to go 100% Haskell with this? Is it realistic to keep some legacy support in pure Python for users who prefer it that way (probably not)?

**Comment by MoritzR** [comment]
Let me quickly answer here.

* `Do you want to go 100% Haskell with this?`
That's not possible because I don't know any other viable alternative to [pyfints](https://github.com/raphaelm/python-fints). This means that the FinTS part will always be written in Python. You can check the remaining python code after switching to haskell [here](https://github.com/MoritzR/fints2ledger/blob/0238799155b5db7dd587195c19215762c000342a/data/pyfints.py). This file basically contains everything needed to download FinTS transactions as CSV. Haskell is used for everything else like config parsing and sorting the transactions into a journal file.

* `Is it realistic to keep some legacy support in pure Python?`
Yes, I want to continue to provide support for any important fixes, like migrating to [pyfints 4](https://python-fints.readthedocs.io/en/latest/upgrading_3_4.html) for example. I won't add features anymore, but let's be real, I also rarely did that on the python version.

* `I am very much willing to contribute further to this project`
That's great to hear 🙂 ! Haskell has a steeper learning curve than Python though, so expect that even smaller changes, like https://github.com/MoritzR/fints2ledger/pull/23 will be noticeably harder. I would also probably need to add at least a bit of documentation and comments to make getting into the codebase easier. Let me know if I should do that.

**Comment by stewmehr** [comment]
> I don't know any other viable alternative to [pyfints](https://github.com/raphaelm/python-fints). This means that the FinTS part will always be written in Python.

That is music to my ears because I have already toyed around a bit with [pyfints](https://github.com/raphaelm/python-fints) and the [mt940 library](https://mt940.readthedocs.io/en/latest/index.html). Sounds like a good entry point for me to find my way around the code base and start working on additional columns/data fields to download.

> Yes, I want to continue to provide support for any important fixes, like migrating to [pyfints 4](https://python-fints.readthedocs.io/en/latest/upgrading_3_4.html) for example.

Sounds great!

> I would also probably need to add at least a bit of documentation and comments to make getting into the codebase easier. Let me know if I should do that.

I do not want to make any promises that I can't keep but my schedule for the foreseeable future looks rather empty. I would like to try and fit learning a new programming language in there. I have been on the road these last few days so I have not been able to check out the Haskell branch and see how much of it I (don't) understand. Still, if you find the time to add some comments/documentation to the existing code I'd highly appreciate that! 🙂

**Comment by MoritzR** [comment]
I have added a few code comments and [some documentation](https://github.com/MoritzR/fints2ledger/blob/master/CONTRIBUTING.md). It's not much but maybe enough to raise additional questions. Feel free to raise an issue regarding any documentation you would like to see. @stewmehr 

Regarding this issue (support for multiple accounts), it's probably not a good first issue. I didn't tackle this for a long time because I dread having to set up some migration for existing users, as the config.yml format would change.

**Comment by stewmehr** [comment]
Thanks a lot, I will start working my way through the codebase and through "Learn You a Haskell for Great Good". As discussed it will ~probably~ definitely be easier for me to start with some `pyfints` related coding.

Slightly OT but since you mentioned a possible migration to `pyfints 4`: do you already know if/how having a fints product ID will be checked and/or enforced in `pyfints 4`? Are developers supposed to request a project-specific product ID that has to be provided in every API call? Are users supposed to request their own? I am a little confused as to how this is supposed to work with open source software.

**Comment by MoritzR** [comment]
> Are users supposed to request their own?

No, I requested a product ID myself, which all users of `fints2ledger` will automatically use.

>  but where do German banks draw the line?

They specifically mention that product IDs of "libraries" should not be used for API requests, see Question 4 on the [ZKA website FAQ](https://www.hbci-zka.de/register/register_faq.htm).

So for now, users of `fints2ledger` should not need to worry about all of that at all.

**Comment by devurandom** [comment]
> I had a look at [buchhaltung](https://github.com/dpaetzel/buchhaltung) and that looks like a much more evolved version of `fints2ledger`. I gave up trying it out after some hours of failing to set it up, though. What's the reason you don't use this over `fints2ledger`?

I just looked into Johannes Gerer's [buchhaltung](https://github.com/johannesgerer/buchhaltung) myself, but it appears it stopped working after PSD2. David Pätzel then rewrote it as [aq2ledger](https://github.com/dpaetzel/aq2ledge) for recent versions of [aqbanking](https://www.aquamaniac.de/rdm/projects/aqbanking), but this was also abandoned 3 years ago due to lack of time.  Thus `fints2ledger` appears to be the only remaining automatic tool.
